---
title: "Applying CTSV to Spatial Transcriptomics Data"
author: "Jingw Yu"
date: "Jan 6th, 2022"
output:  
  html_document:
    keep_md: yes
  rmarkdown::html_vignette:
    keep_md: yes
vignette: >
  %\VignetteIndexEntry{Cell-type-specific spatially variable gene detection}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup,results = 'hide'}
library(CTSV,quietly = TRUE)
library(doSNOW)
library(doParallel)
library(foreach)
```

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache = TRUE,
  out.width = "100%"
)
```
## Introduction
Cell-type-specific Spatially Variable gene detection, or CTSV, is an R package for identifying cell-type-specific spatially variable genes in bulk sptial transcriptomics data. In this Vignette, we will introduce a standard workflow of CTSV. By utilizing single-cell RNA sequencing data as reference, we got cell type proportions in each spot. Then we take the cell type proportion, location coordinates and spatial expression data as input to CTSV.


## Load example data
In order to run RCTD, the first step is to get cell-type proportions. There are some deconvolution methods such as RCTD, SPOTlight and SpatialDWLS. We provide an example data including observed raw count bulk ST data, location coordinate matrix, cell-type proportion matrix and the true SV gene patterns.
```{r ST}
dir <- system.file(package = 'CTSV') #directory for the example data
load(paste0(dir,"/example_data.RData"))
# zero rate in data:
print(summary(rowMeans(Y==0)))
# dimension of bulk ST data
print(dim(Y))
# dimension of cell-type proportion matrix:
print(dim(W))
# SV genes in each cell type:
print(colnames(Y)[which(gamma_true[,1] == 1)])
print(colnames(Y)[which(gamma_true[,2] == 1)])
# Number of SV genes at the aggregated level:
print(sum(rowSums(gamma_true)>0))
```


## Running CTSV
We are now ready to run CTSV on the bulk ST data using `ctsv` function. 

* `Y` is a $n \times G$ count bulk ST data matrix, where each row stands for a spot and each column represents a gene.
* `loc` is a $n \times 2$ location matrix, where each row is a two-dimensional coordinate for the corresponding spot.
* `W` is the cell-type-specific matrix with $n\times K$ dimensions, where $K$ is the number of cell types.
* `num_core:` for parallel processing, the number of cores used. If set to 1, parallel processing is not used. The system will additionally be checked for number of available cores. Note, that we recommend setting `num_core` to at 
least `4` or `8` to improve efficiency.

```{r Run CTSV}
result <- ctsv(Y,loc,W,num_core = 1)
```

## CTSV results

The results of RCTD are located in a list. The first element of 

* `Pval`, combined p-values, a $G\times 2K$ matrix.
* `Qval` stores adjusted q-values of the combined p-values, it is a $G \times 2K$ matrix.

```{r results}
# View on q-value matrix
head(result$qval)
```

Then we want to extra SV genes with an FDR level at 0.05 using `SVGene` function. We regard the q-value matrix return by `CTSVPval` and an threhold of 0.05 as input. The result of the `SVGene` is a list containing two elements, the first of which is a $G\times 2K$ 0-1 matrix indicating SV genes in each cell type and axis.

```{r SVgenes}
re <- SVGene(result$qval,0.05)
# SV genes in each cell type:
print(re$SVGene)
```